src/basic/cgroup-util.c

```diff
@@ -2211,6 +2212,62 @@ int cg_trim_everywhere(CGroupMask supported, const char *path, bool delete_root)
         return 0;
 }
 
+int cg_mask_to_string(CGroupMask mask, char **ret) {
+        _cleanup_strv_free_ char **t = NULL;
+        CGroupController c;
+        char *s;
+
+        assert(ret);
+
+        if (!mask)
+                return 0;
+
+        for (c = 0; c < _CGROUP_CONTROLLER_MAX; c++)
+                if (mask & CGROUP_CONTROLLER_TO_MASK(c)) {
+                        char *n;
+                        int r;
+
+                        n = strdup(cgroup_controller_to_string(c));
+                        if (!n)
+                                return -ENOMEM;
+
+                        r = strv_push(&t, n);
+                        if (r < 0)
+                                return r;
+                }
+
+        s = strv_join(t, NULL);
+        if (!s)
+                return -ENOMEM;
```

Comment:
It appears to me that this does an awful lot of allocations. Any chance you can rework this to make use of GREEDY_REALLOC and directly writing to the resulting string, without the detour of a strv? i.e. for each controller we want to attach, extend a dynamic character array as necessary, and simply add to the end of the string? I figure it'll not just be more efficient but also shorter and prettier.

This function is likely to be called a lot during serialization, hence optimizing this if it's easy sounds like a good idea.